ocs_ci/utility/docker.py

```diff
@@ -0,0 +1,145 @@
+            config.DEPLOYMENT['default_latest_tag'] or 'latest' will be used.
+
+    Returns:
+        str: latest tag for downstream image from docker registry
+
+    Raises:
+        TagNotFoundException: In case no tag found
+
+    """
+    manifests = get_ocs_registry_docker_image_manifests()
+
+    latest_tag = latest_tag or config.DEPLOYMENT.get(
+        'default_latest_tag', 'latest'
+    )
+    latest_image = None
+    for manifest in manifests:
+        tag = manifest['tag']
+        image = manifest['fsLayers']
+        if tag == latest_tag:
+            latest_image = image
+            break
+    if not latest_image:
+        raise TagNotFoundException("Couldn't find latest tag!")
+    latest_tag_found = False
+    for manifest in manifests:
+        tag = manifest['tag']
+        image = manifest['fsLayers']
+        if not upgrade:
+            if (
+                tag not in constants.LATEST_TAGS
+                and image == latest_image
+            ):
+                return tag
+        if upgrade:
+            if not latest_tag_found and tag == latest_tag:
+                latest_tag_found = True
+                continue
+            if latest_tag_found:
+                if (
+                    tag not in constants.LATEST_TAGS
+                    and image != latest_image
+                    and "rc" in tag
+                ):
+                    return tag
+    raise TagNotFoundException("Couldn't find any desired tag!")
+
+
+def get_next_version_available_for_upgrade_docker(current_tag):
+    """
+    This function returns the tag built after the current_version
+
+    Args:
+        current_tag (str): Current build tag from which to search the next one
+            build tag.
+
+    Returns:
+        str: tag for downstream image from docker registry built after
+            the current_tag.
+
+    Raises:
+        TagNotFoundException: In case no tag suitable for upgrade found
+
+    """
+    if current_tag in constants.LATEST_TAGS:
+        return current_tag
+    manifests = get_ocs_registry_docker_image_manifests()
+    current_tag_index = None
+    for index, manifest in enumerate(manifests):
+        if manifest['tag'] == current_tag:
+            if index < 2:
+                raise TagNotFoundException("Couldn't find tag for upgrade!")
+            current_tag_index = index
+            break
+    sliced_reversed_tags = manifests[:current_tag_index]
+    sliced_reversed_tags.reverse()
+    for manifest in sliced_reversed_tags:
+        tag = manifest['tag']
+        if tag not in constants.LATEST_TAGS and "rc" in tag:
+            return tag
+    raise TagNotFoundException("Couldn't find any tag!")
+
+
+def get_ocs_registry_docker_image_manifests():
+    """
+    Retrieves ocs-registry image manifest data for all tags.
+
+    Returns:
+        list: list of manifest dicts
+
+    """
+    # retrieve auth token
+    docker_user = ''  # todo: retrieve from config, probably RUN section
```

Comment:
the credentials we are used to take from ./data folder - do you think that we should try to automatically read some file there so people will not have to always pass the config file?

Reply:
As the token comes from a one-time-viewable window in the docker UI I think we will most likely need to have users create a file and place it in the `data` folder so that we can read the credentials in. This avoids having to require passing a custom config file all the time, but still requires users to do some additional setup if they are going to execute deployments using images from docker. I will go this route and if we are unable to read the credentials I will raise an exception that will point users to the doc describing how to setup their docker credentials.

Reply:
Yep, this sounds good to me. Answered in mail and put all to the infra ticket:
https://projects.engineering.redhat.com/browse/OCSQE-231

Reply:
Credentials are now being pulled from a new file `auth.yaml` in the `data` dir that users will need to create. Doc has been added describing how to set this file up.